ostree-upload-server.py

```diff
@@ -93,54 +63,104 @@ def get_count(self):
         self._webapp_callback = webapp_callback
 
         self.route("/")(self.index)
         self.route("/upload", methods=["GET", "POST"])(self.upload)
+        self.route("/push")(self.push)
 
         self._tempdir = tempfile.mkdtemp(prefix="ostree-upload-server-")
         atexit.register(os.rmdir, self._tempdir)
 
+    def _check_auth(self):
+        if not self._users:
+            return True
+        auth = request.authorization
+        if auth and \
+                auth.username in self._users and \
+                auth.password == self._users.get(auth.username):
+            return True
+        return False
+
+    def _authenticate(self):
+        """Sends a 401 response that enables basic auth"""
+        message = { 'success': False, 'message': "Authentication required" }
+        return Response(
+            json.dumps(message), 401,
+            {'WWW-Authenticate': 'Basic realm="Login Required"'})
+
     def index(self):
         return "<a href='{0}'>upload</a>".format(url_for("upload"))
+        return "<a href='{0}'>push</a>".format(url_for("push"))
 
     def upload(self):
         """
-        Upload a flatpak bundle
+        Receive a flatpak bundle
         """
+        if not self._check_auth():
+            return self._authenticate()
+
         if request.method == "POST":
             logging.debug("/upload: POST request start")
 
-            # TODO: active_upload_counter should be a member
             with self._upload_counter:
                 if 'file' not in request.files:
-                    return "no file in POST\n", 400
+                    message = { 'success': False, 'message': "No file in request" }
+                    return jsonify(message), 400
 
                 upload = request.files['file']
                 if upload.filename == "":
-                    return "no filename in upload\n", 400
+                    message = { 'success': False, 'message': "No filename in request" }
+                    return jsonify(message), 400
 
-                # TODO: tempdir should be a member
                 (f, real_name) = tempfile.mkstemp(dir=self._tempdir)
                 os.close(f)
                 upload.save(real_name)
 
-                # TODO: task_list reference should be a member
-                self._webapp_callback(upload.filename, real_name)
+                self._webapp_callback(ReceiveTask(upload.filename, real_name, self._repo))
                 logging.debug("/upload: POST request completed for " + upload.filename)
 
-                return "task added\n"
+                # TODO: should return a task ID that can be used to check task status
+                message = { 'success': True, 'message': "Importing bundle" }
+                return jsonify(message)
         else:
-            return "only POST method supported\n", 400
+            message = { 'success': False, 'message': "Only POST method is suppported" }
+            return jsonify(message), 400
```

Comment:
Lots of same-template returns of 400 - maybe you could wrap that into a response-returning function?